chain/polkadot/relay_chain_node.go

```diff
@@ -0,0 +1,327 @@
+	return fmt.Sprintf("%s.json", p.Chain.Config().ChainID)
+}
+
+// RawChainSpecFilePathFull returns the full path to the raw chain spec file
+// within the container.
+func (p *RelayChainNode) RawChainSpecFilePathFull() string {
+	return filepath.Join(p.NodeHome(), fmt.Sprintf("%s-raw.json", p.Chain.Config().ChainID))
+}
+
+// RawChainSpecFilePathRelative returns the relative path to the raw chain spec file
+// within the container.
+func (p *RelayChainNode) RawChainSpecFilePathRelative() string {
+	return fmt.Sprintf("%s-raw.json", p.Chain.Config().ChainID)
+}
+
+// GenerateChainSpec builds the chain spec for the configured chain ID.
+func (p *RelayChainNode) GenerateChainSpec(ctx context.Context) error {
+	chainCfg := p.Chain.Config()
+	cmd := []string{
+		chainCfg.Bin,
+		"build-spec",
+		fmt.Sprintf("--chain=%s", chainCfg.ChainID),
+		"--disable-default-bootnode",
+	}
+	res := p.Exec(ctx, cmd, nil)
+	if res.Err != nil {
+		return res.Err
+	}
+	fw := dockerutil.NewFileWriter(p.logger(), p.DockerClient, p.TestName)
+	return fw.WriteFile(ctx, p.VolumeName, p.ChainSpecFilePathContainer(), res.Stdout)
+}
+
+// GenerateChainSpecRaw builds the raw chain spec from the generated chain spec
+// for the configured chain ID.
+func (p *RelayChainNode) GenerateChainSpecRaw(ctx context.Context) error {
+	chainCfg := p.Chain.Config()
+	cmd := []string{
+		chainCfg.Bin,
+		"build-spec",
+		fmt.Sprintf("--chain=%s.json", filepath.Join(p.NodeHome(), chainCfg.ChainID)),
+		"--raw",
+	}
+	res := p.Exec(ctx, cmd, nil)
+	if res.Err != nil {
+		return res.Err
+	}
+	fw := dockerutil.NewFileWriter(p.logger(), p.DockerClient, p.TestName)
+	return fw.WriteFile(ctx, p.VolumeName, p.RawChainSpecFilePathRelative(), res.Stdout)
+}
+
+// CreateNodeContainer assembles a relay chain node docker container ready to launch.
+func (p *RelayChainNode) CreateNodeContainer(ctx context.Context) error {
+	nodeKey, err := p.NodeKey.Raw()
+	if err != nil {
+		return fmt.Errorf("error getting ed25519 node key: %w", err)
+	}
+	multiAddress, err := p.MultiAddress()
+	if err != nil {
+		return err
+	}
+	chainCfg := p.Chain.Config()
+	cmd := []string{
+		chainCfg.Bin,
+		fmt.Sprintf("--chain=%s", p.RawChainSpecFilePathFull()),
+		fmt.Sprintf("--ws-port=%s", strings.Split(wsPort, "/")[0]),
+		fmt.Sprintf("--%s", IndexedName[p.Index]),
+		fmt.Sprintf("--node-key=%s", hex.EncodeToString(nodeKey[0:32])),
+		"--beefy",
+		"--rpc-cors=all",
+		"--unsafe-ws-external",
+		"--unsafe-rpc-external",
+		"--prometheus-external",
+		fmt.Sprintf("--prometheus-port=%s", strings.Split(prometheusPort, "/")[0]),
+		fmt.Sprintf("--listen-addr=/ip4/0.0.0.0/tcp/%s", strings.Split(rpcPort, "/")[0]),
+		fmt.Sprintf("--public-addr=%s", multiAddress),
+		"--base-path", p.NodeHome(),
+	}
+	p.logger().
+		Info("Running command",
+			zap.String("command", strings.Join(cmd, " ")),
+			zap.String("container", p.Name()),
+		)
+
+	cc, err := p.DockerClient.ContainerCreate(
```

Comment:
Is there functionality missing from our `dockerutil.Image` helpers? 

I favor using our docker abstractions before dipping into the lower-level docker client API in other parts of the codebase. Our abstractions will ensure things like proper cleanup. In other words, isolate docker low-level to `dockerutil` pkg and try not to leak it elsewhere. 

If `dockerutil.Image` is missing functionality, we can extend or refactor by offering more config options and perhaps another method for `CreatorContainer()` that creates the container but does not run it. Then, this chain node can hold onto a `dockerutil.Image` instead of a container id.

Reply:
This follows the same strategy for starting the daemon process as the others, `chain_node`, `tendermint_node`, and `penumbra_app_node`.

Reply:
Ah so it is. I feel that's not ideal design and the better one is to depend on our abstractions. In the case of proper cleanup, a new contributor (including our future selves) may not realize the subtleties of tagging, for example. 

There's also quite a bit of docker client duplication with the various chain nodes, so another signal it's time for an abstraction imo. 

But all that's to say; it's probably a future refactor.

Reply:
Yes that would be a great consolidation, I created #237 as a follow-on